<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Chatgpt" data-toc-modified-id="Chatgpt-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Chatgpt</a></span></li><li><span><a href="#Spoonacular-python" data-toc-modified-id="Spoonacular-python-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Spoonacular python</a></span></li><li><span><a href="#Spoonacular-classify" data-toc-modified-id="Spoonacular-classify-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Spoonacular classify</a></span></li><li><span><a href="#Using-dataset" data-toc-modified-id="Using-dataset-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Using dataset</a></span></li></ul></div>

In [ ]:
#77268583b575469e898efe16e4e81960

## Chatgpt

In [1]:
API_KEY = '77268583b575469e898efe16e4e81960'

In [2]:
import requests

# Replace 'YOUR_API_KEY' with your actual Spoonacular API key
BASE_URL = 'https://api.spoonacular.com'

def self_search_recipes_by_ingredients(ingredients, number_of_recipes=5):
    # Construct the search URL
    search_url = f"{BASE_URL}/recipes/findByIngredients"
    
    # Parameters for the API request
    params = {
        'ingredients': ','.join(ingredients),
        'number': number_of_recipes,
        'apiKey': API_KEY
    }
    
    # Make the API request
    response = requests.get(search_url, params=params)
    
    # Check if the request was successful
    if response.status_code == 200:
        recipes = response.json()
        return recipes
    else:
        print(f"Error: {response.status_code}")
        return None


In [5]:
# Example usage
ingredients_list = ['hemp oil cold pressed', 'white toast soft', 'eggs f/range x1 12pk', 'spaghetti rich tomato', '%c 1/2 fam ht chicken']
recipes = self_search_recipes_by_ingredients(ingredients_list)

if recipes:
    for recipe in recipes:
        print(f"Recipe: {recipe['title']}")
        print(f"Used Ingredients: {', '.join([ingredient['name'] for ingredient in recipe['usedIngredients']])}")
        print(f"Missing Ingredients: {', '.join([ingredient['name'] for ingredient in recipe['missedIngredients']])}")
        print(f"URL: https://spoonacular.com/recipes/{recipe['title'].replace(' ', '-')}-{recipe['id']}")
        print()

Recipe: Knishes - Potato Filling
Used Ingredients: chicken fat, egg
Missing Ingredients: onions, potatoes
URL: https://spoonacular.com/recipes/Knishes---Potato-Filling-648983

Recipe: Chicken in a Basil Cream Sauce
Used Ingredients: spaghetti, egg
Missing Ingredients: chicken breasts, panko, parmesan cheese, basil, heavy cream, tomato
URL: https://spoonacular.com/recipes/Chicken-in-a-Basil-Cream-Sauce-638118

Recipe: Extra Large Homemade Meatballs w Pasta
Used Ingredients: egg, spaghetti
Missing Ingredients: ground chuck, garlic, bread, mozzarella, seasoning, onion, pasta/spaghetti sauce
URL: https://spoonacular.com/recipes/Extra-Large-Homemade-Meatballs-w-Pasta-642502

Recipe: Mummy Meatballs with Spaghetti
Used Ingredients: egg, spaghetti noodles
Missing Ingredients: bread crumbs, garlic paste, seasoning, ground beef, parmesan cheese, pimiento stuffed olives, spaghetti sauce, extra wide egg noodles
URL: https://spoonacular.com/recipes/Mummy-Meatballs-with-Spaghetti-652591

Recipe: Le

In [7]:
recipes

[{'id': 633338,
  'title': 'Bacon Wrapped Filet Mignon',
  'image': 'https://img.spoonacular.com/recipes/633338-312x231.jpg',
  'imageType': 'jpg',
  'usedIngredientCount': 1,
  'missedIngredientCount': 2,
  'missedIngredients': [{'id': 10023583,
    'amount': 4.0,
    'unit': 'inches',
    'unitLong': 'inches',
    'unitShort': 'inches',
    'aisle': 'Meat',
    'name': 'beef tenderloin steaks',
    'original': '4 inches beef tenderloin steaks, about 2 thick',
    'originalName': 'beef tenderloin steaks, about 2 thick',
    'meta': ['thick'],
    'image': 'https://img.spoonacular.com/ingredients_100x100/beef-tenderloin.jpg'},
   {'id': 10123,
    'amount': 4.0,
    'unit': 'slices',
    'unitLong': 'slices',
    'unitShort': 'slice',
    'aisle': 'Meat',
    'name': 'bacon',
    'original': '4 slices of smoked bacon',
    'originalName': 'smoked bacon',
    'meta': ['smoked'],
    'extendedName': 'smoked bacon',
    'image': 'https://img.spoonacular.com/ingredients_100x100/raw-bacon.p

In [6]:
# Example usage
ingredients_list = ['chsshrparmesan', 'mush cup', 'cream whipped', 'thickencream', 'garlic']
recipes = self_search_recipes_by_ingredients(ingredients_list)

if recipes:
    for recipe in recipes:
        print(f"Recipe: {recipe['title']}")
        print(f"Used Ingredients: {', '.join([ingredient['name'] for ingredient in recipe['usedIngredients']])}")
        print(f"Missing Ingredients: {', '.join([ingredient['name'] for ingredient in recipe['missedIngredients']])}")
        print(f"URL: https://spoonacular.com/recipes/{recipe['title'].replace(' ', '-')}-{recipe['id']}")
        print()

Recipe: Bacon Wrapped Filet Mignon
Used Ingredients: garlic
Missing Ingredients: beef tenderloin steaks, bacon
URL: https://spoonacular.com/recipes/Bacon-Wrapped-Filet-Mignon-633338

Recipe: Garlicky Kale
Used Ingredients: garlic
Missing Ingredients: balsamic vinegar, curly kale
URL: https://spoonacular.com/recipes/Garlicky-Kale-644387

Recipe: Simple Vinaigrette
Used Ingredients: garlic
Missing Ingredients: dijon mustard, seasoned rice vinegar
URL: https://spoonacular.com/recipes/Simple-Vinaigrette-660158

Recipe: Kale With Red Onion
Used Ingredients: garlic
Missing Ingredients: kale, onion
URL: https://spoonacular.com/recipes/Kale-With-Red-Onion-648729

Recipe: Simple Mashed Yams
Used Ingredients: garlic
Missing Ingredients: cream cheese, yams
URL: https://spoonacular.com/recipes/Simple-Mashed-Yams-651238



In [10]:
# Example usage
ingredients_list = ['persian fetta with chilli cheese']
recipes = self_search_recipes_by_ingredients(ingredients_list)

if recipes:
    for recipe in recipes:
        print(f"Recipe: {recipe['title']}")
        print(f"Used Ingredients: {', '.join([ingredient['name'] for ingredient in recipe['usedIngredients']])}")
        print(f"Missing Ingredients: {', '.join([ingredient['name'] for ingredient in recipe['missedIngredients']])}")
        print(f"URL: https://spoonacular.com/recipes/{recipe['title'].replace(' ', '-')}-{recipe['id']}")
        print()

Recipe: Tomato, Cucumber & Onion Salad with Feta Cheese: Real Convenience Food
Used Ingredients: persian cucumber
Missing Ingredients: balsamic vinegar, feta cheese, grape tomatoes, onion
URL: https://spoonacular.com/recipes/Tomato,-Cucumber-&-Onion-Salad-with-Feta-Cheese:-Real-Convenience-Food-716416

Recipe: Rice-less Spicy Tuna Hand Rolls
Used Ingredients: persian cucumber
Missing Ingredients: pickled ginger, reduced fat mayonnaise, reduced sodium soy sauce, scallion into pieces, sriracha chili sauce, sushi nori, sashimi grade tuna
URL: https://spoonacular.com/recipes/Rice-less-Spicy-Tuna-Hand-Rolls-658295

Recipe: Winter Fattoush Salad
Used Ingredients: persian cucumber
Missing Ingredients: agave nectar, feta cheese, flat leaf parsley leaves, garlic, ground sumac, lacinato kale leaves, lemon juice, lentils, onion, parsley, pita breads, radishes, savoy cabbage, za'atar
URL: https://spoonacular.com/recipes/Winter-Fattoush-Salad-665378

Recipe: Spiced Chicken With Black Bean and Mango

In [13]:
# Example usage
ingredients_list = ['white toast soft']
recipes = self_search_recipes_by_ingredients(ingredients_list)

if recipes:
    for recipe in recipes:
        print(f"Recipe: {recipe['title']}")
        print(f"Used Ingredients: {', '.join([ingredient['name'] for ingredient in recipe['usedIngredients']])}")
        print(f"Missing Ingredients: {', '.join([ingredient['name'] for ingredient in recipe['missedIngredients']])}")
        print(f"URL: https://spoonacular.com/recipes/{recipe['title'].replace(' ', '-')}-{recipe['id']}")
        print()

Recipe: Kale Bruschetta
Used Ingredients: bread
Missing Ingredients: garlic clove, kale
URL: https://spoonacular.com/recipes/Kale-Bruschetta-648715

Recipe: Cranberry Pecan Greek Yogurt Chicken Salad
Used Ingredients: chipotle fried egg avocado toast
Missing Ingredients: this healthy cranberry pecan greek yogurt chicken salad is easy, planet roasted chicken breast, scallion, cranberries, pecans, combine all, easy black bean soup, easy avocado egg quinoa breakfast bowl
URL: https://spoonacular.com/recipes/Cranberry-Pecan-Greek-Yogurt-Chicken-Salad-879449



In [14]:
# Example usage
ingredients_list = ['butter chicken simmer sauce mild']
recipes = self_search_recipes_by_ingredients(ingredients_list)

if recipes:
    for recipe in recipes:
        print(f"Recipe: {recipe['title']}")
        print(f"Used Ingredients: {', '.join([ingredient['name'] for ingredient in recipe['usedIngredients']])}")
        print(f"Missing Ingredients: {', '.join([ingredient['name'] for ingredient in recipe['missedIngredients']])}")
        print(f"URL: https://spoonacular.com/recipes/{recipe['title'].replace(' ', '-')}-{recipe['id']}")
        print()

Recipe: Bunny Eggs Kids Breakfast
Used Ingredients: mild cheddar cheese
Missing Ingredients: bacon, blueberries, eggs
URL: https://spoonacular.com/recipes/Bunny-Eggs-Kids-Breakfast-636458

Recipe: Oven-Baked Feta Cheese Dip
Used Ingredients: mild cheddar cheese
Missing Ingredients: feta cheese, garlic, tomato sauce, part-skim mozzarella cheese
URL: https://spoonacular.com/recipes/Oven-Baked-Feta-Cheese-Dip-654241

Recipe: Patio Daddio Queso Dip
Used Ingredients: mexican mild
Missing Ingredients: chili beans, new mexico chili powder, chorizo, hot sauce, milk
URL: https://spoonacular.com/recipes/Patio-Daddio-Queso-Dip-655015

Recipe: Baked Macaroni and Cheese
Used Ingredients: mild cheddar
Missing Ingredients: mozzarella, milk, butter, parsley to garnish, elbow macaroni
URL: https://spoonacular.com/recipes/Baked-Macaroni-and-Cheese-1487863

Recipe: Tantalizing White Onion & Tuna Tart
Used Ingredients: ground mild cheese
Missing Ingredients: ready-made puff pastry, butter, onions, eggs, c

In [5]:
# Example usage
ingredients_list = ['chicken', 'bread', 'garlic']
recipes = self_search_recipes_by_ingredients(ingredients_list)

if recipes:
    for recipe in recipes:
        print(f"Recipe: {recipe['title']}")
        print(f"Used Ingredients: {', '.join([ingredient['name'] for ingredient in recipe['usedIngredients']])}")
        print(f"Missing Ingredients: {', '.join([ingredient['name'] for ingredient in recipe['missedIngredients']])}")
        print(f"URL: https://spoonacular.com/recipes/{recipe['title'].replace(' ', '-')}-{recipe['id']}")
        print()

Recipe: Karahi Chicken
Used Ingredients: chicken, garlic cloves, naan bread/chapatis to serve
Missing Ingredients: bay leaf, chicken stock, coriander, garam masala, ghee, ground coriander seeds, mint, onion
URL: https://spoonacular.com/recipes/Karahi-Chicken-648746

Recipe: Chili Lime Chicken Burgers
Used Ingredients: chicken burgers, garlic cloves, buns
Missing Ingredients: ground chicken, onion, bell pepper, cilantro, pepper flakes, chili powder, lime juice, guacamole, avocados, cumin, bell pepper
URL: https://spoonacular.com/recipes/Chili-Lime-Chicken-Burgers-638566

Recipe: Kale Bruschetta
Used Ingredients: bread, garlic clove
Missing Ingredients: kale
URL: https://spoonacular.com/recipes/Kale-Bruschetta-648715

Recipe: Lemon and Garlic Slow Roasted Chicken
Used Ingredients: chicken weighing 2.3kg, bulbs garlic
Missing Ingredients: lemons
URL: https://spoonacular.com/recipes/Lemon-and-Garlic-Slow-Roasted-Chicken-649495

Recipe: Gambas Al Ajo
Used Ingredients: slc bread, clv garlic


## Spoonacular python

In [ ]:
#https://pypi.org/project/spoonacular/

In [3]:
import spoonacular as sp
api = sp.API(API_KEY)

In [8]:
# Parse an ingredient
response = api.parse_ingredients("Woolworths White Toast Soft 65Cy", servings=1)
data = response.json()
print(data[0]['name'])

woolworths toast 65cy


In [9]:
data

[{'id': 18070,
  'original': 'Woolworths White Toast Soft 65Cy',
  'originalName': 'Woolworths White Toast Soft 65Cy',
  'name': 'woolworths toast 65cy',
  'amount': 1.0,
  'unit': 'serving',
  'unitShort': 'serving',
  'unitLong': 'serving',
  'possibleUnits': ['slice thin', 'piece', 'slice', 'g', 'ounce', 'oz', 'cup'],
  'estimatedCost': {'value': 0.36, 'unit': 'US Cents'},
  'consistency': 'solid',
  'shoppingListUnits': ['ounces', 'pounds'],
  'aisle': 'Bakery/Bread',
  'image': 'toast.jpg',
  'meta': ['white', 'soft']}]

In [11]:
# Get a random food joke
response = api.get_a_random_food_joke()
data = response.json()
print(data['text'])

Sometimes when I reflect back on all the beer I drink I feel shamed. Then I look into the glass and think about the workers in the brewery and all of their hopes and dreams. If I didn't drink this beer, they might be out of work and their dreams would be shattered. Then I say to myself, "It is better that I drink this beer and let their dreams come true than be selfish and worry about my liver.


In [5]:
ingredients_list = ['chicken', 'tomato', 'garlic']

In [6]:
# Get a random food joke
response = api.search_recipes_by_ingredients(ingredients = ingredients_list)
data = response.json()
print(data)

[{'id': 640803, 'title': 'Crispy Buttermilk Fried Chicken', 'image': 'https://img.spoonacular.com/recipes/640803-312x231.jpg', 'imageType': 'jpg', 'usedIngredientCount': 1, 'missedIngredientCount': 2, 'missedIngredients': [{'id': 2028, 'amount': 1.0, 'unit': 'teaspoon', 'unitLong': 'teaspoon', 'unitShort': 'tsp', 'aisle': 'Spices and Seasonings', 'name': 'paprika', 'original': '1 teaspoon paprika', 'originalName': 'paprika', 'meta': [], 'image': 'https://img.spoonacular.com/ingredients_100x100/paprika.jpg'}, {'id': 1230, 'amount': 1.0, 'unit': 'cup', 'unitLong': 'cup', 'unitShort': 'cup', 'aisle': 'Milk, Eggs, Other Dairy', 'name': 'buttermilk', 'original': '1 cup buttermilk', 'originalName': 'buttermilk', 'meta': [], 'image': 'https://img.spoonacular.com/ingredients_100x100/buttermilk.jpg'}], 'usedIngredients': [{'id': 5006, 'amount': 2.0, 'unit': 'pounds', 'unitLong': 'pounds', 'unitShort': 'lb', 'aisle': 'Meat', 'name': 'chicken', 'original': '2 pounds chicken', 'originalName': 'chi

In [9]:
data

[{'id': 640803,
  'title': 'Crispy Buttermilk Fried Chicken',
  'image': 'https://img.spoonacular.com/recipes/640803-312x231.jpg',
  'imageType': 'jpg',
  'usedIngredientCount': 1,
  'missedIngredientCount': 2,
  'missedIngredients': [{'id': 2028,
    'amount': 1.0,
    'unit': 'teaspoon',
    'unitLong': 'teaspoon',
    'unitShort': 'tsp',
    'aisle': 'Spices and Seasonings',
    'name': 'paprika',
    'original': '1 teaspoon paprika',
    'originalName': 'paprika',
    'meta': [],
    'image': 'https://img.spoonacular.com/ingredients_100x100/paprika.jpg'},
   {'id': 1230,
    'amount': 1.0,
    'unit': 'cup',
    'unitLong': 'cup',
    'unitShort': 'cup',
    'aisle': 'Milk, Eggs, Other Dairy',
    'name': 'buttermilk',
    'original': '1 cup buttermilk',
    'originalName': 'buttermilk',
    'meta': [],
    'image': 'https://img.spoonacular.com/ingredients_100x100/buttermilk.jpg'}],
  'usedIngredients': [{'id': 5006,
    'amount': 2.0,
    'unit': 'pounds',
    'unitLong': 'pounds'

In [29]:
def get_recipe_details(ids, includeNutrition=None):
        """ Get information about a recipe.
            https://spoonacular.com/food-api/docs#get-recipe-information
        """
        endpoint = "recipes/{ids}/information".format(ids=ids)
        url_query = {}
        url_params = {"includeNutrition": includeNutrition}
        return requests.get(endpoint, method="GET", query_=url_query, params_=url_params)
    

get_recipe_information_1(ids = 1230)

response = api.get_analyzed_recipe_instructions("'Crispy Buttermilk Fried Chicken'")
data = response.json()
print(data)

In [34]:
#final recipe information. 
response = api.get_recipe_information(640803)
data = response.json()
data

{'vegetarian': False,
 'vegan': False,
 'glutenFree': False,
 'dairyFree': False,
 'veryHealthy': False,
 'cheap': False,
 'veryPopular': False,
 'sustainable': False,
 'lowFodmap': False,
 'weightWatcherSmartPoints': 11,
 'gaps': 'no',
 'preparationMinutes': None,
 'cookingMinutes': None,
 'aggregateLikes': 53,
 'healthScore': 4,
 'creditsText': 'foodista.com',
 'sourceName': 'foodista.com',
 'pricePerServing': 66.93,
 'extendedIngredients': [{'id': 5006,
   'aisle': 'Meat',
   'image': 'whole-chicken.jpg',
   'consistency': 'SOLID',
   'name': 'chicken',
   'nameClean': 'whole chicken',
   'original': '2 pounds chicken',
   'originalName': 'chicken',
   'amount': 2.0,
   'unit': 'pounds',
   'meta': [],
   'measures': {'us': {'amount': 2.0, 'unitShort': 'lb', 'unitLong': 'pounds'},
    'metric': {'amount': 435.449, 'unitShort': 'g', 'unitLong': 'grams'}}},
  {'id': 20081,
   'aisle': 'Baking',
   'image': 'flour.png',
   'consistency': 'SOLID',
   'name': 'flour',
   'nameClean': 'wh

In [35]:
# Get a random food joke
response = api.detect_food_in_text("Woolworths White Toast Soft 65Cy")
data = response.json()
data

{'annotations': [{'annotation': 'toast',
   'tag': 'dish',
   'image': 'https://img.spoonacular.com/ingredients_100x100/toast'}],
 'processedInMs': 7}

## Spoonacular classify

In [ ]:
def classify_a_grocery_product(self, product):
        """ Given a grocery product title, this endpoint allows
            you to detect what basic ingredient it is.
            https://spoonacular.com/food-api/docs#classify-a-grocery-product
        """
        endpoint = "food/products/classify"
        url_json = product
        quota = {'requests': 0, 'tinyrequests': 1, 'results': 0}
        return self._make_request(endpoint, method="POST", json_=url_json)

## Using dataset

In [37]:
import pandas as pd

In [39]:
recipe_data = pd.read_csv(r"C:\Users\rishi\OneDrive - Monash University\Documents\Monash\MDS Y2 S2\IE\Iteration 2\recipe_Data\dataset\full_dataset.csv")

In [ ]:
recipe_data = recipe_data.drop(["Unnamed: 0", "link", "source"], axis = 1)

In [41]:
recipe_data.shape

(2231142, 7)

In [44]:
recipe_data.loc[0, "NER"]

'["brown sugar", "milk", "vanilla", "nuts", "butter", "bite size shredded rice biscuits"]'

In [45]:
recipe_data.loc[0, "directions"]

'["In a heavy 2-quart saucepan, mix brown sugar, nuts, evaporated milk and butter or margarine.", "Stir over medium heat until mixture bubbles all over top.", "Boil and stir 5 minutes more. Take off heat.", "Stir in vanilla and cereal; mix well.", "Using 2 teaspoons, drop and shape into 30 clusters on wax paper.", "Let stand until firm, about 30 minutes."]'

In [51]:
recipe_data.columns

Index(['title', 'ingredients', 'directions', 'NER'], dtype='object')

In [52]:
recipe_data.head(2)

title                                        ingredients  \
0    No-Bake Nut Cookies  ["1 c. firmly packed brown sugar", "1/2 c. eva...   
1  Jewell Ball'S Chicken  ["1 small jar chipped beef, cut up", "4 boned ...   

                                          directions  \
0  ["In a heavy 2-quart saucepan, mix brown sugar...   
1  ["Place chipped beef on bottom of baking dish....   

                                                 NER  
0  ["brown sugar", "milk", "vanilla", "nuts", "bu...  
1  ["beef", "chicken breasts", "cream of mushroom...

I want to generate recipes based on NER
e.g.
- I have "milk", "vanilla", "chicken", "rice"
- give top 5 matches that has most of these ingredients
- show the ones that are used and not used in the recipe. 
- then when one of them is clicked --> the instructions is shown

Ideas:
- ML model. x = NER. y = Title ?
- decision tree model like thing - recipes under brown sugar. another branch of milk. now get the branch that goes down atleast a bit. 
- Fuzzy match - match your list with that NER list
- for loop. get input[i] in NER. get recipes with max count on input in NER.